src/main/java/com/google/sps/servlets/GetBlurAreasServlet.java

```diff
@@ -0,0 +1,211 @@
+
+      for (AnnotateImageResponse res : responses) {
+        if (res.hasError()) {
+          continue;
+        }
+
+        // For each face detected create a rectangle represented by an array of points.
+        for (FaceAnnotation face : res.getFaceAnnotationsList()) {
+          ArrayList<Point> facePoly = new ArrayList<>();
+          for (Vertex vertex : face.getFdBoundingPoly().getVerticesList())
+            facePoly.add(new Point(vertex.getX(), vertex.getY()));
+
+          // Add the rectangle to our list of face rectangles.
+          facePolys.add(facePoly);
+        }
+      }
+
+      client.close();
+    }
+
+    return facePolys;
+  }
+
+  /**
+   * Returns the BlobKey that points to the file uploaded by the user, or null if the user didn't
+   * upload a file.
+   */
+  private BlobKey getBlobKey(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so the BlobKey is empty. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    return blobKey;
+  }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the image represented by the
+   * binary data stored at the BlobKey parameter.
+   */
+  private byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // End index is inclusive, so we have to subtract 1 to get fetchSize bytes.
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // If we read fewer bytes than we requested, then we reached the end.
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+
+  private Boolean isImageSupported(BlobKey blobKey) {
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    return supportedTypes.contains(blobInfo.getContentType());
```

Comment:
So it doesn't check the extension of the file, right? It has some other mechanism to figure out the type of the content? If so, you should fix the error message to the user about the extension.

Reply:
Right, it doesn't check for the extension, it checks for its type. I edited the message, thanks for pointing out!